src/tool/hpcrun/gpu/intel/opencl-activity-translate.c

```diff
@@ -0,0 +1,106 @@
+
+#include "opencl-activity-translate.h"
+#include "opencl-api.h" //profilingData
+
+void convert_kernel_launch(gpu_activity_t *, void *, cl_event);
+void convert_memcpy(gpu_activity_t *, void *, cl_event, gpu_memcpy_type_t);
+profilingData* getTimingInfoFromClEvent(cl_event);
+profilingData* getMemoryProfileInfo(profilingData*, cl_memory_callback*);
+gpu_kernel_t openclKernelDataToGenericKernelData(struct profilingData *);
+gpu_memcpy_t openclMemDataToGenericMemData(struct profilingData *);
+
+void opencl_activity_translate(gpu_activity_t * ga, cl_event event, void * user_data)
+{
+  cl_generic_callback* cb_data = (cl_generic_callback*)user_data;
+  opencl_call type = cb_data->type;
+  switch (type)
+  {
+	case kernel:
+	  convert_kernel_launch(ga, user_data, event);
+	  break;
+	case memcpy_H2D:
+	  convert_memcpy(ga, user_data, event, GPU_MEMCPY_H2D);
+	  break;
+	case memcpy_D2H:
+	  convert_memcpy(ga, user_data, event, GPU_MEMCPY_D2H);
+	  break;
+  }
+  cstack_ptr_set(&(ga->next), 0);
+}
+
+void convert_kernel_launch(gpu_activity_t * ga, void * user_data, cl_event event)
+{
+  cl_kernel_callback* kernel_cb_data = (cl_kernel_callback*)user_data;
+  ETMSG(OPENCL, "Saving kernel data to gpu_activity_t");
+  ga->kind = GPU_ACTIVITY_KERNEL;
+  profilingData* pd = getTimingInfoFromClEvent(event);
+  ga->details.kernel = openclKernelDataToGenericKernelData(pd);
+  set_gpu_interval(&ga->details.interval, pd->startTime, pd->endTime);
+  ga->details.kernel.correlation_id = kernel_cb_data->correlation_id;
+  free(pd);
+}
+
+void convert_memcpy(gpu_activity_t * ga, void * user_data, cl_event event, gpu_memcpy_type_t kind)
+{
+  cl_memory_callback* memory_cb_data = (cl_memory_callback*)user_data;
+  profilingData* pd = getTimingInfoFromClEvent(event);
+  pd = getMemoryProfileInfo(pd, memory_cb_data);
+  ga->kind = GPU_ACTIVITY_MEMCPY;
+  ga->details.memcpy = openclMemDataToGenericMemData(pd);
+  ga->details.memcpy.correlation_id = memory_cb_data->correlation_id;
+  set_gpu_interval(&ga->details.interval, pd->startTime, pd->endTime);
+  free(pd);
+}
+
+// can we remove extra time variables?
+profilingData* getTimingInfoFromClEvent(cl_event event)
+{
+  cl_ulong commandQueued = 0;
+  cl_ulong commandSubmit = 0;
+  cl_ulong commandStart = 0;
+  cl_ulong commandEnd = 0;
+  cl_int errorCode = CL_SUCCESS;
+
+  errorCode |= clGetEventProfilingInfo(event, CL_PROFILING_COMMAND_QUEUED, sizeof(commandQueued), &commandQueued, NULL);
+  errorCode |= clGetEventProfilingInfo(event, CL_PROFILING_COMMAND_SUBMIT, sizeof(commandSubmit), &commandSubmit, NULL);
+  errorCode |= clGetEventProfilingInfo(event, CL_PROFILING_COMMAND_START, sizeof(commandStart), &commandStart, NULL);
+  errorCode |= clGetEventProfilingInfo(event, CL_PROFILING_COMMAND_END, sizeof(commandEnd), &commandEnd, NULL);
+  if (errorCode != CL_SUCCESS)
+	ETMSG(OPENCL, "error in collecting profiling data");
+  profilingData *pd = (profilingData*) malloc(sizeof(profilingData));
```

Comment:
I don't recommend that we allocate and free one data structure per opencl callback

Reply:
done